bids/reports/utils.py

```diff
@@ -0,0 +1,573 @@
+            if len(run_dict[scan]) > 1:
+                s = 's'
+            else:
+                s = ''
+            run_str = list_to_str(run_dict[scan])
+            string = '{rs} run{s} of the {sc}'.format(rs=run_str,
+                                                      s=s,
+                                                      sc=scan)
+            out_list.append(string)
+        for_str = ' for the {0}'.format(list_to_str(out_list))
+    else:
+        for_str = ''
+
+    desc = '''
+           A {variants} {seqs} field map (phase encoding:
+           {dir_}; {n_slices} slices; repetition time, TR={tr}ms;
+           echo time, TE={te}ms; flip angle, FA={fa}<deg>;
+           field of view, FOV={fov}mm; matrix size={ms};
+           voxel size={vs}mm) was acquired{for_str}.
+           '''.format(variants=variants,
+                      seqs=seqs,
+                      dir_=dir_,
+                      for_str=for_str,
+                      n_slices=n_slices,
+                      tr=num_to_str(metadata['RepetitionTime']*1000),
+                      te=num_to_str(metadata['EchoTime']*1000),
+                      fa=metadata['FlipAngle'],
+                      vs=vs_str,
+                      fov=fov_str,
+                      ms=ms_str)
+    desc = desc.replace('\n', ' ').lstrip()
+    while '  ' in desc:
+        desc = desc.replace('  ', ' ')
+
+    return desc
+
+
+def final_paragraph(metadata):
+    """
+    Describes dicom-to-nifti conversion process and methods generation.
+
+    Parameters
+    ----------
+    metadata : :obj:`dict`
+        The metadata for the scan.
+
+    Returns
+    -------
+    desc : :obj:`str`
+        Output string with scanner information.
+    """
+    if 'ConversionSoftware' in metadata.keys():
+        software_str = ' using {soft} ({conv_vers})'.format(soft=metadata['ConversionSoftware'],
+                                                            conv_vers=metadata['ConversionSoftwareVersion'])
+    else:
+        software_str = ''
+    desc = '''
+           Dicoms were converted to NIfTI-1 format{software_str}.
+           This section was (in part) generated
+           automatically using pybids ({meth_vers}).
+           '''.format(software_str=software_str,
+                      meth_vers=__version__)
+    desc = desc.replace('\n', ' ').lstrip()
+    while '  ' in desc:
+        desc = desc.replace('  ', ' ')
+
+    return desc
+
+
+def parse_niftis(layout, niftis, subj, ses, config):
+    """
+    Loop through niftis in a BIDSLayout and generate the appropriate description
+    type for each scan. Compile all of the descriptions into a list.
+
+    Parameters
+    ----------
+    layout : :obj:`bids.grabbids.BIDSLayout`
+        Layout object for a BIDS dataset.
+    niftis : :obj:`list` or :obj:`grabbit.core.File`
+        List of nifti files in layout corresponding to subject/session combo.
+    subj : :obj:`str`
+        Subject ID.
+    ses : :obj:`str` or :obj:`None`
```

Comment:
If this can be `None`, set that as the default value in the argument list. This should also make your life easier when handling sessions as I suggested above.